internal/comment/cache.go

```diff
@@ -0,0 +1,35 @@
+package comment
+
+import (
+	"go/ast"
+	"go/token"
+	"sync"
+)
+
+type Cache struct {
+	comments map[*ast.File]ast.CommentMap
```

Comment:
Possible optimizations I've realized when looking more into analysis framework: I believe analysis drivers like golangci-lint will never visit the same file more than once during an execution. So:

1. it's not necessary to cache the comment map at all per file; it can be stored as a single variable in some pass-level visitor object
2. there's no need to force atomic operations with mutexes in this context

More generally, it should be possible to limit the use of mutexes by using package-specific caches and only writing to the cache of the pass' own package (the package-to-cache lookup would still need to be synchronized due to concurrent package processing). There is no risk of concurrent read from and write to the same package cache because every reasonable driver will sequentially process packages in topological order.

Conclusion: it should only be necessary to do a single mutex lock when defining the package-specific cache in the cache lookup. After that, you can preserve the pointer to this package-specific cache in the visitor and use zero synchronization.

Unclear what perf impacts would be in practice. I might try this and see.

Reply:
1) performance impact is very small.
2) this cache is also expected to be used for definition directives, which are expected to be hit more than once during analysis.